src/aktualizr_primary/secondary.cc

```diff
@@ -14,18 +15,19 @@
 namespace Primary {
 
 using Secondaries = std::vector<std::shared_ptr<Uptane::SecondaryInterface>>;
-using SecondaryFactoryRegistry = std::unordered_map<std::string, std::function<Secondaries(const SecondaryConfig&)>>;
+using SecondaryFactoryRegistry =
+    std::unordered_map<std::string, std::function<Secondaries(const SecondaryConfig&, Aktualizr& aktualizr)>>;
 
-static Secondaries createIPSecondaries(const IPSecondariesConfig& config);
+static Secondaries createIPSecondaries(const IPSecondariesConfig& config, Aktualizr& aktualizr);
 
 static SecondaryFactoryRegistry sec_factory_registry = {
     {IPSecondariesConfig::Type,
-     [](const SecondaryConfig& config) {
+     [](const SecondaryConfig& config, Aktualizr& aktualizr) {
        auto ip_sec_cgf = dynamic_cast<const IPSecondariesConfig&>(config);
-       return createIPSecondaries(ip_sec_cgf);
+       return createIPSecondaries(ip_sec_cgf, aktualizr);
```

Comment:
Passing in the Aktualizr object here seems like a big change, but I don't see any immediate concerns. Are there other ways to do this that would simplify this?

Reply:
Maybe the new methods to Aktualizr is the biggest change in this context. We're coming from `initSecondaries` that already takes an `Aktualizr` instance. But these are the two sides of the same problem

The alternatives that I know of would be:
- to pass the storage directly in some way (but it's opaque from here, as it's in Aktualizr). Having a `Aktualizr::GetStorage()` would be quite ugly.
- create another storage instance from the storage path and use it there. Probably not SQLStorage but something that derives from SQLStorageBase. I am would be worried about the added boilerplate.
- use another dedicated storage mechanism (?)

Reply:
Back then when I introduced this 'Primary::initSecondaries' my intention/plan was and still is to consider it as an initial version of generic Abstract Factory that is part of Secondary Registry mechanism that we might introduce in the future. So, I consider 'Primary::initSecondaries' as something that belongs to Primary/main namespace and it's OK to pass an instance of Aktualizr to it. Having said that, I think that it's not OK to pass this instance to a specific secondary factory method, like createIPSecondaries() as it should belong IpUptaneSecondary namespace. Also,  createIPSecondaries() should be under IpUptaneSecondary namespace too, I did not put it there in the first place because we didn't want to spend much time to implementation of some secondary registration mechanism (e.g. something similar to what Laurent proposes for the package manager registry).

Reply:
It's probably fine as is, but my concern is that the Aktualizr class is mostly meant to be an API for library users, and I think we're extending that a bit here. I fear that something similar is happening to what has happened to SotaUptaneClient. I'm not sure how to fix that, and it might just take some time to step back and figure out a good refactoring strategy. Or maybe I'm worrying too much about things that don't matter.

Reply:
> It's probably fine as is, but my concern is that the Aktualizr class is mostly meant to be an API for library users, and I think we're extending that a bit here.

I share your concern but in this case, adding a secondary implementation is expected by our library users, so this might be in the scope. But it's indeed not clear cut.

Reply:
IMHO, both Aktualizr::GetStorage() and std::vector<SecondaryInfo> Aktualizr::GetSecondaries() along with Aktualizr::SetSecondaryData() are quite ugly. Also, I have the same feeling about adding of SotaUptaneClient::waitSecondariesReachable() into SotaUptaneClient. 
IMHO, both of them are consequences of trying to use business logic specific to one secondary (IP secondary in our case) or a few secondaries (but not all, probable future case, e.g. IP Secondary, CAN Secondary but not Virtual Secondary or Docker Secondary) in an object/objects that operate over array of secondaries (Aktualizr, SotaUptaneClient).

Even, if some business logic,e.g. this "wait" stuff, is applicable/relevant to some of the secondaries it SHOULD NOT be injected into Aktualizr/SotaUptaneClient but just simply reused in the relevant secondaries' implementation by means of composition (e.g. the way we reuse the uptane verification logic) or inheritance (base class that include the wait logic). In any case, this logic should be "hidden" behind the secondary interface.

That's why I strongly against of Aktualizr::GetStorage(), Aktualizr::GetSecondaries(), Aktualizr::SetSecondaryData(), SotaUptaneClient::waitSecondariesReachable() as well as this really ugly SecondaryInterface::ping() which is BTW also another consequances of injecting secondary specific business logic into SotaUptaneClient.

I think if we just reuse SQLStorageBase() in createIPSecondaries then we would not need all the aforementioned functions at all and overall implementation will be actually simpler compared to the current implementation.

In the long run, I think it absolutely makes sense to reconsider the storage/persistence engine for secondary because of the usage of a SQL engine for storing a relatively small number of records in a single table seam like way overhead. BTW, I think, the same is true for libaktualizr, why do we need SQL DB for it? Why not use some simple a key-value storage, e.g. https://github.com/google/leveldb (this is just example).

Reply:
> In the long run, I think it absolutely makes sense to reconsider the storage/persistence engine for secondary because of the usage of a SQL engine for storing a relatively small number of records in a single table seam like way overhead. BTW, I think, the same is true for libaktualizr, why do we need SQL DB for it? Why not use some simple a key-value storage, e.g. [google/leveldb](https://github.com/google/leveldb) (this is just example).

This is another discussion entirely and frankly it sounds like change for the sake of change to me. I believe the main advantages we got from using sqlite was to have atomic transactions and corruption resistance (leveldb seems to have some type of transactions FWIW) and it's also a very largely deployed and tested piece of code with few dependencies. We haven't had a problem with the supposed overhead that we pay now but we'd definitely have issues migrating to something else.

Reply:
> Even, if some business logic,e.g. this "wait" stuff, is applicable/relevant to some of the secondaries it SHOULD NOT be injected into Aktualizr/SotaUptaneClient but just simply reused in the relevant secondaries' implementation by means of composition (e.g. the way we reuse the uptane verification logic) or inheritance (base class that include the wait logic). In any case, this logic should be "hidden" behind the secondary interface.

While I agree that the composition or inheritance options are viable, I disagree that the `wait` or `ping` logic is unique to IP Secondaries. These are methods that we will probably need for almost every Secondary that sits on a separate system, which is the default expectation. Virtual and Docker Secondaries are an exception to the rule, and arguably still not the definitive solution to the problem we are using them to solve. I'd rather treat the expected use case as the default and optimize for that. We can revise the current solution later, but I think it's acceptable as is.

> In the long run, I think it absolutely makes sense to reconsider the storage/persistence engine for secondary because of the usage of a SQL engine for storing a relatively small number of records in a single table seam like way overhead.

I agree, we probably need to rework this sooner or later. Feel free to make a ticket to track that work.

> the same is true for libaktualizr, why do we need SQL DB for it? Why not use some simple a key-value storage, e.g. https://github.com/google/leveldb (this is just example).

SQLite is pretty small and has some nice atomicity guarantees. Does leveldb really provide the same features as SQLite? How easy would it be to support migration, table joins, typed columns, etc.? I'm open to alternative storage backends but I don't want to invest the effort without a substantial reason to believe that it would be a significant improvement.

Reply:
> I disagree that the `wait` or `ping` logic is unique to IP Secondaries.

It's not exactly what I say. I mean, I am not saying that `wait` or `ping` logic is unique to IP Secondaries, I think and agree that it can be relevant for more than one secondary. What I am trying to say is that this logic can be reused/applied in another way, e.g. just include it into implementations of secondaries that need it, what IMHO, simpler, cleaner,  compared to adding it to sotauptaneclient/aktualizr and polluting interface.

Reply:
> a very largely deployed and tested piece of code with few dependencies. We haven't had a problem with the supposed overhead that we pay now but we'd definitely have issues migrating to something else

Yes, I agree, sqlite has been proven by the time and is quite reliable piece of software and there is no particular reason for replacing it, at the moment. Yet, SQL DB usage in case of aktualizr's use-case is quite overhead and dealing with the stuff that Patrick mentioned would be much simpler.

Reply:
> What I am trying to say is that this logic can be reused/applied in another way, e.g. just include it into implementations of secondaries that need it, what IMHO, simpler, cleaner, compared to adding it to sotauptaneclient/aktualizr and polluting interface.

I think I see your point now. Do you agree that we need some abstract notion of `wait` and `ping`? The implementations can indeed live in the subclasses, but my point was that we need some way to ask if a Secondary is available.

Reply:
@patrickvacek Mike suggested that if we want to wait for secondaries, we could so in the implementation of `install()` for example. I don't think he advocates having abstract notions of `wait()` and `ping()` as virtual methods (@mike-sul please correct me if I'm wrong).

In principle I think it's a good design and I also agree that it's dangerous to make this interface grow too much but there are things that would be harder to implement this way:

- waiting a fixed time for all secondaries globally and not waiting N times
- do not attempt to update the primary if secondaries are not available and the updates target them. We can still rollback and that's maybe something we should think about anyway (since secondary install can still fail even we found the secondaries)

We might rethink these things or just decide that we don't need it but I think that the clean solution is also less powerful.

Reply:
> I don't think he advocates having abstract notions of `wait()` and `ping()` as virtual methods (@mike-sul please correct me if I'm wrong).

that's correct. I strongly against it :).

> Do you agree that we need some abstract notion of wait and ping? The implementations can indeed live in the subclasses, but my point was that we need some way to ask if a Secondary is available.

No, I disagree. IMHO, we already have functionality that kind of "waits" for IP Secondary, it's all software stack running under Asn1Rpc(). What we need is just extend/wrap this functionality in order to add additional policy(ies) into it, e.g. desired timeout/wait time as in this specific case  and then use parametrized Asn1Rpc() from IpUptaneSecondary methods. For example, in this specific case, we could call Asn1Rpc(timeout=10m) in IpUptaneSecondary::install() and Asn1Rpc(timeout=60s) in another (but this is also questionable since we might wanna wait "longer" at the stage of sending metadata too).

Reply:
The solution you're describing just doesn't have the same behavior as what we've agreed on when we took on this task. The two points I've raised are what we want to satisfy right now. 

I sympathize with you as you want to keep away the bloat and coupling and I see that this change can seem to go in the wrong direction and might make things more difficult to untangle later.

However after discussing with @patrickvacek , we still think the best way forward right now is to try to get this merged first. The two main reasons are:

- to implement what we agreed on right now, we believe that overriding the `install()` step is not sufficient. As we are giving more fine-grained control on the installation process, we will inevitably need more fine-grained interfaces. It does not necessarily have to be something low-level like `ping()`, we could also imagine having a `globalPreInstall()` step in `SecondaryInterface()` but that would maybe require more careful design, that we haven't yet started
- we do not claim this is a perfect solution that will be set in stone. Though, moving the piece of code later when we have a more satisfying design should be relatively easy. There is a balance to be found between a hacky solution and an all-encompassing plan that is never implemented in practice. And as we are trying to work in an iterative fashion, the current version feels like it's just good enough right now.

Reply:
> just doesn't have the same behavior as what we've agreed on when we took on this task

It won't be exactly the same internally, but effect/outcome from an external standpoint of view will be the same.

> overriding the install() step is not sufficient.

I don't think we need to override install() step, we were mistaken referring to install() method at all, it should be rather sendFirmware() or putMetadata() or both.

> we could also imagine having a globalPreInstall() step in SecondaryInterface() but that would maybe require more careful design, that we haven't yet started

I am not sure if I understand if we need to have it at all. I think this idea about "waiting" for secondaries prior to sending firmware to it is conceptually flawed if to look into it from a broader perspective and take into account other aspects such as layered nature on networking stack and potential use-case of "dynamic" secondary support.

> There is a balance to be found between a hacky solution and an all-encompassing plan that is never implemented in practice. And as we are trying to work in an iterative fashion, the current version feels like it's just good enough right now.

That's fair enough and makes sense in many cases but in this particular case the implementation option that goes in line with an all-encompassing plan (proper design) is simpler and easier compared to the given implementation.

@lbonn @patrickvacek There is no need to persuade me :). We've had a great and constructive discussion around this PR and it's absolutely OK that different people have different opinions. I fully understand that discussions cannot last forever, and a decision should be made. What worked the best for me at my previous companies if there was no full consensus among peers is to let a tech lead/architect (or owner of software under discussion) to make a final decision and everyone else should accept and complain to it.

So, that's absolutely fine with me if this change is merged :) as long as this change fulfill the original goal/issue and the CI jobs are happy.

Reply:
I appreciate the robust discussion; I think that leads us to a better solution in the end and helps keep us aware of the trade-offs and alternatives. My suggestion for now is to move forward with this PR as is to satisfy our immediate product goals, and then have a followup discussion (perhaps at the sprint review or the next relevant grooming) to consider the longer-term product goals and then decide where to go from here.